### PR TITLE
Release v2.1.0

### PR DESCRIPTION
**Pull request by rschouten97**
Updated to latest best practices
Moved permissions such as InSite, OutSite, Profit Windows, etc. to permissions script instead of setting these within the account lifecycle actions, e.g. enable or disable

### CODE REVIEW

**Comment by Rick-Jongbloed** [comment]
Versioning table needs updating

**Comment by Rick-Jongbloed** [comment]
I think its better to use the mapping of the AD, not Azure AD. We should keep both (employee and user) connectors in line with each other

**Comment by Rick-Jongbloed** [comment]
createUser in settings != createAccount in createscript.
Solution: change settings to createAccount

**Comment by Rick-Jongbloed** [comment]
enable actie, $updateConnector -> $c.updateConnector

**Comment by Rick-Jongbloed** [comment]
Not tested update script as i didn't need it at this client

**Comment by Rick-Jongbloed** [comment]
The message "No changes needed" can be omitted in the enable and disable steps if there are no changes _and_ an enable or disable action is executed and audit log for a enable/disable is already shown.

**Comment by Rick-Jongbloed** [comment]
Error when setting Outsite on an existing acocunt, where there is no emailadress filled in.

 14:51:42.21 Error granting permission [OutSite] to AFAS user [91307.003068]. Error Message: Profit kan de gebruiker niet opslaan omdat geen e-mailadres is ingevuld. Het e-maildres is verplicht voor gebruikers die toegang hebben tot een portal. Vink het veld portal uit of voer het e-mailadres van de gebruiker in en probeer opnieuw.

**Comment by rhouthuijzen** [comment]
> enable actie, $updateConnector -> $c.updateConnector

Not only in enable but also in disable and delete

**Comment by rhouthuijzen** [comment]
If you add mutiple permissions to one person and you grant or revoke them on the same time you have a high change on getting errors like listed below. (concurent actions disabled)


 11:55:44.51 Error enabling AFAS user [x]. Error Message: De gegevens zijn inmiddels gewijzigd door een andere gebruiker.
 11:55:44.51 Failed to enable account , message: Error running script

 11:58:14.92 Error granting permission [Profit Windows] to AFAS user [x].. Error Message: De gegevens zijn inmiddels gewijzigd door een andere gebruiker.
 11:58:14.92 Failed to grant permission '{ "PermissionDefinitionId": "4fc6bf80-d407-46a7-bf74-e14100b62441", "PermissionIdentification": { "Name": "Profit Windows", "Id": "Awin" } }', message: Error running script

 11:58:14.92 Error granting permission [InSite] to AFAS user [x]. Error Message: De gegevens zijn inmiddels gewijzigd door een andere gebruiker.
 11:58:14.92 Failed to grant permission '{ "PermissionDefinitionId": "4fc6bf80-d407-46a7-bf74-e14100b62441", "PermissionIdentification": { "Name": "InSite", "Id": "InSi" } }', message: Error running script

11:58:14.41 Error granting permission [AFAS Accept] to AFAS user [x]. Error Message: De gegevens zijn inmiddels gewijzigd door een andere gebruiker.
 11:58:14.41 Failed to grant permission '{ "PermissionDefinitionId": "4fc6bf80-d407-46a7-bf74-e14100b62441", "PermissionIdentification": { "Id": "AcUs", "Name": "AFAS Accept" } }', message: Error running script

**Comment by rschouten97** [comment]
@Rick-Jongbloed and @rhouthuijzen all the comments and change requests have been impmented.
In future, please use the review option in Github. That way you can point to a specific line or section code.
And we can subsequently address the approval comments once we have implemented the necessary changes.

File: grantPermission.ps1 [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    # Grant permission
+    try {
+        $bodyAddPermission = [PSCustomObject]@{
+            'KnUser' = @{
+                'Element' = @{
+                    # Gebruiker
+                    '@UsId'  = $currentAccount.Gebruiker
+                    'Fields' = @{
+                        # Mutatie code
+                        'MtCd'   = $account.MtCd
```
Comment (mouki9) [inline grantPermission.ps1:179]: $account.MtCd should be 1

**Review by mouki9 — COMMENTED** [review]
bug in grantPermission.ps1

File: grantPermission.ps1 (hunk 1 above)
Comment (rhouthuijzen) [reply grantPermission.ps1]: Permissions need to be further tested before we add this to the connector

File: README.md [hunk 2]
```diff
@@ -1,17 +1,24 @@
...
 </p>
 
 ## Versioning
-| Version | Description | Date |
-| - | - | - |
-| 2.0.0   | Release of v2 connector including performance and logging upgrades | 2022/08/30  |
-| 1.0.0   | Initial release | 2020/07/24  |
+| Version | Description                                                        | Date       |
+| ------- | ------------------------------------------------------------------ | ---------- |
+| 2.0.0   | Release of v2 connector including performance and logging upgrades | 2022/08/30 |
```
[the remark this answers is not in the document]
Comment (rhouthuijzen) [reply README.md]: Permissions need to be further tested before we add this to the connector

File: README.md [hunk 3]
```diff
@@ -36,13 +43,16 @@ The interface to communicate with Profit is through a set of GetConnectors, whic
...
+| Action                             | Action(s) Performed                                | Comment |
+| ---------------------------------- | -------------------------------------------------- | ------- |
+| create.ps1                         | Create or Update AFAS user                         | Optionally you can update the UserID to match the mapped convention. **Note that this is not advised as this can break certain links in AFAS. Use with care!**. Users are only created or updated when this is configured, **make sure to check your configuration options to prevent unwanted actions**.                                                                           |
+| enable.ps1                         | Unblock and optinally update AFAS user             | Optionally, you can provide additional properties to update, e.g. **"EmAd", "Upn"**. The **default example sets these with the AD values**. This action is perfomed with **entry code "6"**. For more information on the entry codes, see the [AFAS documenation](https://help.afas.nl/help/en/SE/App_Conect_UpdDsc_KnUser.htm).                                                    |
```
[the remark this answers is not in the document]
Comment (rhouthuijzen) [reply README.md:42]: Fixed in commit

File: README.md [hunk 4]
```diff
@@ -36,13 +43,16 @@ The interface to communicate with Profit is through a set of GetConnectors, whic
...
+| update.ps1                         | Update AFAS user                                   | Update with the specified properties, e.g. **"EmAd", "Upn"**, etc. The **default example sets these with the AD values**. This action is perfomed with **entry code "1"**. For more information on the entry codes, see the [AFAS documenation](https://help.afas.nl/help/en/SE/App_Conect_UpdDsc_KnUser.htm)                                                                       |
+| disable.ps1                        | Block and optinally update AFAS user               | Optionally, you can provide additional properties to update, e.g. **"EmAd", "Upn"**.  The **default example clears these values**, as the values have to be unique over all AFAS environments. This action is perfomed with **entry code "2"**. For more information on the entry codes, see the [AFAS documenation](https://help.afas.nl/help/en/SE/App_Conect_UpdDsc_KnUser.htm). |
```
[the remark this answers is not in the document]
Comment (rhouthuijzen) [reply README.md:44]: fixed in commit

File: create.ps1 [hunk 5]
```diff
@@ -158,420 +124,464 @@ function Resolve-AFASErrorMessage {
...
-                            Message = "Successfully updated AFAS account with userId '$($aRef.Gebruiker)'. (No Changes needed)"
-                            IsError = $false
-                        })
-                }
-                else {
-                    Write-Warning "DryRun: No changes to AFAS account with userId '$($currentAccount.Gebruiker)'"
-                }
-                break
+        
+            # Define ExportData with account fields and correlation property
```
[the remark this answers is not in the document]
Comment (rhouthuijzen) [reply create.ps1:532]: Will be fixed in PSv2

File: create.ps1 [hunk 6]
```diff
@@ -158,420 +124,464 @@ function Resolve-AFASErrorMessage {
...
-                        IsError = $false
+                        # Action  = "" # Optional
+                        Message = "Error creating AFAS user [$($account.UsId)]. Error Message: $($errorMessage.AuditErrorMessage). Account object: $($createAccount | ConvertTo-Json -Depth 10)"
+                        IsError = $true
                     })
             }
-            else {
-                Write-Warning "DryRun: Would create AFAS account with userId '$($account.'KnUser'.'Element'.'@UsId')', Business Email: '$($account.'KnUser'.'Element'.'Fields'.'EmAd')' and UPN: '$($account.'KnUser'.'Element'.'Fields'.'Upn')'"
+            
+            # Define ExportData with account fields and correlation property
```
[the remark this answers is not in the document]
Comment (rhouthuijzen) [reply create.ps1:322]: This part is not in the catch. The bracket ends on line 320

**Review by rhouthuijzen — APPROVED** [review]
Good changes especially the mapping and the differences between old and new values

Additional information:
Permissions need to be further tested before we add this to the connector.

**Pull request merged**